File: backend/src/Maran.Modules/Ftp/Domain/Entities/FtpUser.cs
namespace Maran.Modules.Ftp.Domain.Entities;

/// <summary>
/// One FTPS login created for an account: a real system account on the host, jailed into that
/// account's own FTPS jail with the account's real home bind-mounted inside it (spec §11). This row
/// — not <c>/etc/passwd</c> — is the record of who owns what.
/// </summary>
/// <remarks>
/// <para>
/// <b>This entity is the tenant boundary of every customer-facing path in this module.</b> The
/// host's user namespace is global and the host has no notion of a tenant, so a login name only
/// looks like it belongs to an account because of the prefix the panel put there. Deciding ownership
/// from those names is a prefix scan, and a prefix scan aliases account <c>alice</c> onto
/// <c>alice_bob</c>'s logins — <c>alice_bob_deploy</c> starts with <c>alice_</c>. Every listing,
/// read, delete and password reset is therefore authorised by these rows through the context's
/// tenant query filter. The agent holds the same line one layer down and by a different mechanism:
/// it matches the candidate's passwd home against the account's jail rather than decomposing its
/// name (<c>ops/src/ftps/delete_ftps_user.rs</c>).
/// </para>
/// <para>
/// <b>There is no password column, of any kind — not plaintext, not encrypted, not a hash.</b> The
/// panel mints a password, shows it once and forgets it; the host's own shadow entry is the only
/// copy in the system. A stored copy, however it is encrypted, is a copy that can be read out of the
/// panel's database, and a panel that can read back every customer's FTPS password is a single theft
/// away from every customer's files. A customer who lost theirs gets a new one set
/// (<c>ResetFtpUserPassword</c>), never the old one shown again. That absence is asserted by a
/// reflection test rather than left to a reviewer noticing a new property one day, because it is the
/// axis on which this design silently stops being true.
/// </para>
/// <para>
/// <b>There is no jail path here and there is nothing for one to come from.</b> The agent derives
/// the jail from the validated account name alone and creates it root-owned, so the customer names
/// no directory: there is nothing to validate, nothing to store, and the chroot-escape class of bug
/// has nothing to aim at.
/// </para>
/// <para>
/// <b>There is no protocol column either, and its absence is the opposite of the DTO's presence.</b>
/// Every row in this table is an FTPS login by construction — the module owns one daemon — so a
/// column could hold only one value. <c>FtpUserDto</c> nevertheless carries the label, because the
/// SPA merges this module's logins with the Sftp module's on one screen and must render a fact the
/// backend produced rather than one derived from which URL it happened to call.
/// </para>
/// <para>
/// Nothing here has a public setter and there is no method that changes any of it, because nothing
/// about a login changes after it is made: its home, jail, shell and group are all derived from the
/// account rather than chosen, renaming a system user is a delete and a create, and the password —
/// the one thing that does change — is deliberately not here.
/// </para>
/// </remarks>
public sealed class FtpUser
{
    /// <summary>The row's identity, and the only identifier a customer's request may name.</summary>
    public Guid Id { get; private set; }

    /// <summary>The account that owns this login. Every tenant-scoped query is closed over this column.</summary>
    public Guid AccountId { get; private set; }

    /// <summary>The name the customer asked for, without the account prefix.</summary>
    /// <remarks>
    /// What the customer typed and what a screen shows them, so a customer never has to know the
    /// prefix exists. Unique within the account and NOT across the host: the prefix is what makes
    /// <c>files</c> available to every account at once.
    /// </remarks>
    public string Name { get; private set; }

    /// <summary>The system login as the host actually holds it, as the agent reported it.</summary>
    /// <remarks>
    /// Recorded rather than rebuilt from <see cref="Name"/> and the account's user name. Rebuilding
    /// would make this row's truth depend on the panel and the agent agreeing about a separator
    /// forever, and the day they disagreed the panel would show a customer a login name they cannot
    /// sign in with.
    /// </remarks>
    public string FullName { get; private set; }

    /// <summary>The instant the login was created.</summary>
    public DateTimeOffset CreatedAt { get; private set; }

    /// <summary>Records an FTPS login the agent has already created on the host.</summary>
    /// <param name="id">The row's identity.</param>
    /// <param name="accountId">The account that owns this login.</param>
    /// <param name="name">The name the customer asked for, without the account prefix.</param>
    /// <param name="fullName">The system login the host holds, as the agent reported it.</param>
    /// <param name="createdAt">The creation instant, taken from <see cref="IClock"/>.</param>
    public FtpUser(Guid id, Guid accountId, string name, string fullName, DateTimeOffset createdAt)
    {
        Id = id;
        AccountId = accountId;
        Name = name;
        FullName = fullName;
        CreatedAt = createdAt;
    }

    /// <summary>Parameterless constructor required by EF Core materialization.</summary>
    private FtpUser()
    {
        Name = string.Empty;
        FullName = string.Empty;
    }
}

File: frontend/src/types/ftpUser.ts
/**
 * The customer FTPS login domain: the row, the request that creates one, and the two responses that
 * carry a password exactly once.
 *
 * It mirrors the backend's `Ftp` module DTOs field-for-field, the way `sftpUser.ts` mirrors the
 * `Sftp` module's. The two stay separate types for the same reason the two modules stay separate on
 * the server; what merges them is `types/fileTransferLogin.ts`, in the SPA, where the merge belongs.
 */

/**
 * Outward view of one customer FTPS login, mirroring the backend's `FtpUserDto`.
 *
 * There is no password here and there is nothing for one to come from: the `Ftp` module holds no
 * password column of any kind, plaintext or hashed. The value exists in exactly two responses — the
 * one that created the login and the one that reset it — and in the host's shadow file.
 */
export interface FtpUser {
  /** The login's identity, and the only identifier a request may name. */
  id: string
  /** The account that owns this login. */
  accountId: string
  /** The name the customer asked for, WITHOUT the account prefix. */
  name: string
  /**
   * The system login the host holds — `<account>_<name>`, as it appears in `/etc/passwd`.
   *
   * This is what the customer types into their FTPS client.
   */
  fullName: string
  /**
   * Which daemon the backend says accepts this login, as the server spells it.
   *
   * It is on the wire rather than assumed by the SPA because the merged screen renders a
   * backend-supplied fact, not one derived from which URL was called. The value is narrowed to
   * `types/fileTransferLogin.ts`'s union at the store boundary; a token this bundle does not know is
   * refused there rather than rendered.
   */
  protocol: string
  /** The instant the login was created, as an ISO-8601 string. */
  createdAt: string
}

/**
 * Request body for `POST /api/v1/ftp-users`, binding the backend's `CreateFtpUserCommand`.
 *
 * No password field and no chroot path: the panel mints the credential, and the jail is derived from
 * the account by the agent — so there is no directory for a request to name and none to be trusted
 * with. `ftp.proto` reserves the field name for exactly this reason.
 */
export interface CreateFtpUserRequest {
  /** The account that will own the login. */
  accountId: string
  /** The login name, without the account prefix; lowercase letters and digits only. */
  name: string
}

/**
 * What `POST /api/v1/ftp-users` answered, mirroring the backend's `CreatedFtpUserDto` — one of the
 * only two responses that ever carry an FTPS password.
 */
export interface CreatedFtpUser {
  /** The new login's identity. */
  id: string
  /** The account that owns it. */
  accountId: string
  /** The name the customer asked for, without the account prefix. */
  name: string
  /** The system login the host holds — what the customer signs in with. */
  fullName: string
  /** Which daemon accepts it, as the server spells it. */
  protocol: string
  /**
   * The generated password. Nothing keeps a copy — not this SPA, not the panel, not the agent — so
   * it is shown once and recovered only by setting a new one.
   */
  password: string
  /** The instant the login was created, as an ISO-8601 string. */
  createdAt: string
}

/**
 * What `POST /api/v1/ftp-users/{id}/password` answered, mirroring `FtpUserPasswordDto`. The only
 * recovery path a lost password has.
 */
export interface FtpUserPassword {
  /** The login that was re-credentialled. */
  id: string
  /** The system login the new password belongs to. */
  fullName: string
  /** The new password, shown once and stored nowhere. */
  password: string
}

/**
 * Typed access to the FTPS users endpoints. Called from Pinia stores only (rules/vue.md).
 */
export interface FtpUsersApi {
  /**
   * Lists the FTPS logins the caller may see.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The logins, in the order the panel reports them.
   */
  list: (signal?: AbortSignal) => Promise<FtpUser[]>

  /**
   * Creates an FTPS login.
   * @param request The owning account and the name the customer chose.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The login as created, including the password shown once.
   */
  create: (request: CreateFtpUserRequest, signal?: AbortSignal) => Promise<CreatedFtpUser>

  /**
   * Gives the login a new password. A login belonging to somebody else answers 404, never 403.
   * @param id The login to re-credential.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The login and its new password, shown once.
   */
  resetPassword: (id: string, signal?: AbortSignal) => Promise<FtpUserPassword>

  /**
   * Removes the login, and only the login: the account's files stay where they are.
   * @param id The login to remove.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns Whether the panel removed the login.
   */
  remove: (id: string, signal?: AbortSignal) => Promise<boolean>
}

/**
 * An FTPS credential the panel is showing for the only time it ever will.
 *
 * It carries the login and the password and NOT the connection facts. The host name and the control
 * port live on `FtpsStatus`, which `GET /api/v1/ftps-server` serves to administrators only, so a
 * customer's dialog has no way to learn them — and the one thing it must never do is compose a host
 * name of its own from `window.location`, because the server's certificate is issued for the
 * panel's own hostname and any other name produces a mismatch warning. The dialog therefore renders
 * the facts when the status is readable and says plainly that it cannot when it is not.
 */
export interface RevealedFtpCredential {
  /** The system login the password belongs to, prefixed exactly as the host holds it. */
  fullName: string
  /** The generated password, exactly as the panel sent it. */
  password: string
}
